Add Learn PWA: Part 18, Window management. (#7721)
GoogleChrome/web.dev: src/site/content/en/learn/pwa/windows/index.md

1- Your PWA gets the token, stores the token, and renders the PWA.

#### Forcing a browser's navigation

If you want to force open the browser with a URL and not an in-app browser, you can use the `_blank` target of `<a href>` elements. This works only on desktop PWAs; on mobile devices, there is no option to open a browser with a URL.

```js
function openBrowser(url) {
	window.open("url", "_blank", "");
}
```

### Opening new windows

On desktop, users can open more than one window of the same PWA. Each window will be a different navigation to the same `start_url`, as if you were opening two browser tabs of the same URL.
From the menu in the PWA, users can select File then New window, and from your PWA code, you can open a new window with the `open()` function. Check the [documentation](https://developer.mozilla.org/docs/Web/API/Window/open) for details.

```js
function openNewWindow() {
	window.open("/", "new-window", "width=600,height=600");
}
```
{% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/CURym4RD66CqcTw4iPl5.png", alt="The same installed PWA with several windows opened on a desktop operating system.", width="800", height="580" %}

> Calling `open()` within a PWA window on iOS or iPadOS returns `null` and doesn't open a window. Opening new windows on Android creates a new in-app browser for the URL—even if the URL is within the scope of your PWA—that typically doesn't trigger an external browsing experience.

Reviewer: ```suggestion

Calling `open()` within a PWA window on iOS or iPadOS returns `null` and doesn't open a window. Opening new windows on Android creates a new in-app browser for the URL—even if the URL is within the scope of your PWA—that typically doesn't trigger an external browsing experience.
```

{% Aside %}
In the [Experimental chapter](/learn/pwa/experimental), you will see how you can start using experimental capabilities.
{% endAside %}

### Windows control overlay

We've mentioned that you can change the window's title by defining the value of the `<title>` element or the `document.title` property. But it's always a string value. What if we could design the title bar as we wish, with HTML, CSS, and images?
That's where Windows Control Overlay comes in, a new experimental capability in Microsoft Edge and Google Chrome for desktop PWAs.

{% Aside %}
Tabbed mode and windows control overlay capabilities define new values for the manifest's `display` member. To create compatibility with all devices, the manifest group has created a way to replace the [fallback `display` chain](/learn/pwa/app-design/#display-modes), so you can specify what `display` value to use if your first option is not available. Read more about it at [Preparing for the display modes of tomorrow](/display-override/).
{% endAside %}

You can read more about this capability at [Customize the window controls overlay of your PWA's title bar](/window-controls-overlay/).

> {% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/9azCEwKziNBBxrTOn3pn.avif", alt="With windows control overlay, you can render content in the title bar.", width="800", height="544" %}

Reviewer: ```suggestion
{% Img src="image/kheDArv5csY6rvQUJDbWRscckLr1/B5UDAJzKTqCRdSy6B5is.png", alt="With windows control overlay, you can render content in the title bar.", width="800", height="544" %}
```

---

> layout: post

Reviewer: I don't think we use this here

      window.resizeTo(800, 600);
   }
});
```

You can query the current screen size and position using the `window.screen` object; you can detect when the window is resized using the `resize` event from the `window` object. There is no event for capturing the window move, so your option is to query the position frequently.

{% Aside %}
Instead of moving and resizing the window absolutely, you can move relatively and resize using `moveBy()` and `resizeBy()`.
{% endAside %}

{% Aside %}
On mobile devices, moving or resizing your window won't make any changes on the screen.
{% endAside %}

### Browsing to other sites
If you want to send the user to an external site that is out of the scope of your PWA, you can do so with a standard `<a href>` HTML element, using `location.href` or opening a new window on compatible platforms.

Currently on all browsers, if your PWA is installed, when you browse to a URL that is out of the [scope of your manifest](/learn/pwa/web-app-manifest/#recommended-fields), the browser engine of your PWA will render an in-app browser within the context of your window.

Some features of the in-app browsers are:

* They appear on top of your content.

> * They have a static URL bar showing the current origin, the window's title, and a menu. Typically, they'reit's themed with the `theme_color` of your manifest.

Reviewer: ```suggestion
* They have a static URL bar showing the current origin, the window's title, and a menu. Typically, it's themed with the `theme_color` of your manifest.
```

* From the contextual menu, you can open that URL in the browser.
* Users can close the browser or go back.

> {% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/0SsThiNz4GDkJy6OiTHj.png", alt="An in-app browser on a desktop PWA when browsing a URL that is out-of-scope", width="800", height="497" %}

Reviewer: ```suggestion
{% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/0SsThiNz4GDkJy6OiTHj.png", alt="An in-app browser on a desktop PWA when browsing a URL that is out-of-scope.", width="800", height="497" %}
```

{% Aside %}
While the in-app browser is on the screen, your PWA is waiting in the background, as if another window is obscuring it.
{% endAside %}

{% Aside 'gotchas' %}
On iOS and iPadOS, the in-app browser uses SafariViewController, a Safari-rendering engine isolated from your PWA storage. Therefore, if the user has a session in Safari, it won't appear in your PWA, but it will appear within an in-app browser in your PWA.
{% endAside %}

> {% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/Gr4qJUx6hAFMyPLPyfQY.png", alt="An in-app browser on an iPhone when browsing a URL that is out-of-scope within a standalone PWA", width="800", height="370" %}

Reviewer: ```suggestion
{% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/Gr4qJUx6hAFMyPLPyfQY.png", alt="An in-app browser on an iPhone when browsing a URL that is out-of-scope within a standalone PWA.", width="800", height="370" %}
```

> {% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/UyohW4R5Sy79ffHri8cz.png", alt="An in-app browser on Android when browsing a URL that is out-of-scope within a standalone PWA", width="800", height="1644" %}

Reviewer: ```suggestion
{% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/UyohW4R5Sy79ffHri8cz.png", alt="An in-app browser on Android when browsing a URL that is out-of-scope within a standalone PWA.", width="800", height="1644" %}
```

1- The in-app browser closes itself when it detects a URL that is part of the scope of the PWA.
1- The engine redirects the main PWA window navigation to the URL that the authentication server went to while in the in-app browser.
1- Your PWA gets the token, stores the token, and renders the PWA.

#### Forcing a browser's navigation

If you want to force open the browser with a URL and not an in-app browser, you can use the `_blank` target of `<a href>` elements. This works only on desktop PWAs; on mobile devices, there is no option to open a browser with a URL.

```js
function openBrowser(url) {
	window.open("url", "_blank", "");
}
```

### Opening new windows

On desktop, users can open more than one window of the same PWA. Each window will be a different navigation to the same `start_url`, as if you were opening two browser tabs of the same URL.
From the menu in the PWA, users can select File then New window, and from your PWA code, you can open a new window with the `open()` function. Check the [documentation](https://developer.mozilla.org/docs/Web/API/Window/open) for details.

```js
function openNewWindow() {
	window.open("/", "new-window", "width=600,height=600");
}
```

> {% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/CURym4RD66CqcTw4iPl5.png", alt="The same installed PWA with several windows opened on a desktop operating system", width="800", height="580" %}

Reviewer: ```suggestion
{% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/CURym4RD66CqcTw4iPl5.png", alt="The same installed PWA with several windows opened on a desktop operating system.", width="800", height="580" %}
```

* Dynamically changing the `document.title` string property at any time.

On desktop PWAs, the title is essential, and it's used in the title bar of the window and sometimes in the task manager or multi-task selection. If you have a single-page application, you may want to update your title on every route.

{% Aside %}
To reduce phishing within PWAs, some desktop browsers may have additional measures on the window's title. For example, if you don't use your app's name in the title, the browser may add a prefix with your PWA's name on it. Other browsers may render the current origin in the title bar for a few seconds when you change the title to highlight where the user navigated.
{% endAside %}

### Tabbed mode

An experimental capability, known as *tabbed mode*​ will let your PWA have a tab-based design similar to a web browser. In this case, the user can have several tabs opened from the same PWA but all tied together in the same operating system window, as you can see in the following image:

<video autoplay="" loop="" muted=""><source src="https://storage.googleapis.com/web-dev-uploads/video/8WbTDNrhLsU0El80frMBGE4eMCD3/JwN0btyXFGiT9oPFh2qJ.webm" type="video/webm"></video>

> You can read more about this experimental capability at (Tabbed application mode for PWA [/tabbed-application-mode/]

Reviewer: ```suggestion
You can read more about this experimental capability at [Tabbed application mode for PWA](/tabbed-application-mode/).
```

{% Aside %}
In the [Experimental chapter](/learn/pwa/experimental), you will see how you can start using experimental capabilities.
{% endAside %}

### Windows control overlay

We've mentioned that you can change the window's title by defining the value of the `<title>` element or the `document.title` property. But it's always a string value. What if we could design the title bar as we wish, with HTML, CSS, and images?
That's where Windows Control Overlay comes in, a new experimental capability in Microsoft Edge and Google Chrome for desktop PWAs.

{% Aside %}
Tabbed mode and windows control overlay capabilities define new values for the manifest's `display` member. To create compatibility with all devices, the manifest group has created a way to replace the [fallback `display` chain](/learn/pwa/app-design/#display-modes), so you can specify what `display` value to use if your first option is not available. Read more about it at [Preparing for the display modes of tomorrow](/display-override/).
{% endAside %}

You can read more about this capability at [Customize the window controls overlay of your PWA's title bar](/window-controls-overlay/).

> {% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/9azCEwKziNBBxrTOn3pn.avif", alt="With windows control overlay, you can render content in the title bar", width="800", height="544" %}

Reviewer: here's the avif

Author: I have no way to get that.

Reviewer: I fixed it :)

> {% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/9azCEwKziNBBxrTOn3pn.avif", alt="With windows control overlay, you can render content in the title bar", width="800", height="544" %}

Reviewer: ```suggestion
{% Img src="image/RK2djpBgopg9kzCyJbUSjhEGmnw1/9azCEwKziNBBxrTOn3pn.avif", alt="With windows control overlay, you can render content in the title bar.", width="800", height="544" %}
```
